cmd/kubeapps-apis/plugins/fluxv2/packages/v1alpha1/cache.go

```diff
@@ -0,0 +1,324 @@
+	// and .Add() is expected to be done by the unit test client. The server-side only signals
+	// when indexing one repo is complete
+	indexRepoWaitGroup *sync.WaitGroup
+}
+
+func newCache(clientGetter server.KubernetesClientGetter) (*fluxPlugInCache, error) {
+	log.Infof("+newCache")
+	REDIS_ADDR, ok := os.LookupEnv("REDIS_ADDR")
+	if !ok {
+		return nil, status.Errorf(codes.FailedPrecondition, "missing environment variable REDIS_ADDR")
+	}
+	REDIS_PASSWORD, ok := os.LookupEnv("REDIS_PASSWORD")
+	if !ok {
+		return nil, status.Errorf(codes.FailedPrecondition, "missing environment variable REDIS_PASSWORD")
+	}
+	REDIS_DB, ok := os.LookupEnv("REDIS_DB")
+	if !ok {
+		return nil, status.Errorf(codes.FailedPrecondition, "missing environment variable REDIS_DB")
+	}
+
+	REDIS_DB_NUM, err := strconv.Atoi(REDIS_DB)
+	if err != nil {
+		return nil, err
+	}
+
+	log.Infof("newCache: addr: [%s], password: [%s], DB=[%d]", REDIS_ADDR, REDIS_PASSWORD, REDIS_DB)
+
+	return newCacheWithRedisClient(
+		clientGetter,
+		redis.NewClient(&redis.Options{
+			Addr:     REDIS_ADDR,
+			Password: REDIS_PASSWORD,
+			DB:       REDIS_DB_NUM,
+		}))
+}
+
+func newCacheWithRedisClient(clientGetter server.KubernetesClientGetter, redisCli *redis.Client) (*fluxPlugInCache, error) {
+	log.Infof("+newCacheWithRedisClient")
+
+	if redisCli == nil {
+		return nil, status.Errorf(codes.FailedPrecondition, "server not configured with redis Client")
+	}
+
+	if clientGetter == nil {
+		return nil, status.Errorf(codes.FailedPrecondition, "server not configured with configGetter")
+	}
+
+	// sanity check that the redis client is connected
+	pong, err := redisCli.Ping(redisCli.Context()).Result()
+	if err != nil {
+		return nil, err
+	}
+	log.Infof("[PING] -> [%s]", pong)
+
+	c := fluxPlugInCache{
+		watcherStarted: false,
+		watcherMutex:   sync.Mutex{},
+		redisCli:       redisCli,
+		clientGetter:   clientGetter,
+	}
+	go c.startHelmRepositoryWatcher()
+	return &c, nil
+}
+
+func (c *fluxPlugInCache) startHelmRepositoryWatcher() {
+	log.Infof("+fluxv2 startHelmRepositoryWatcher")
+	c.watcherMutex.Lock()
+	// can't defer c.watcherMutex.Unlock() because when all is well,
+	// we never return from this func
+
+	if !c.watcherStarted {
+		ch, err := c.newHelmRepositoryWatcherChan()
+		if err != nil {
+			c.watcherMutex.Unlock()
+			log.Errorf("failed to start HelmRepository watcher due to: %v", err)
+			return
+		}
+		c.watcherStarted = true
+		c.watcherMutex.Unlock()
+		log.Infof("watcher successfully started. waiting for events...")
+
+		c.processEvents(ch)
+	} else {
+		c.watcherMutex.Unlock()
+		log.Infof("watcher already started. exiting...")
+	}
+	// we should never reach here
+	log.Warningf("-fluxv2 startHelmRepositoryWatcher")
+}
+
+func (c *fluxPlugInCache) newHelmRepositoryWatcherChan() (<-chan watch.Event, error) {
+	// TODO this is a temp hack to get around the fact that the only clientGetter we've got today
+	// always expects authorization Bearer token in the context
```

Comment:
It always expects the bearer token if it's not setup to use the service account. Right, so we need to ensure that the client can also be used with an explicit request to use the service's service account right?

Reply:
Not quite. I am going to say this my testing shows that the bearer token is ALWAYS expected with the current state of client getter
https://github.com/kubeapps/kubeapps/blob/master/cmd/kubeapps-apis/server/plugins.go#L307
I could be wrong, but the call to extractToken() is the very first thing that happens and that's what fails in my case. If I am right, as I said before this is going to take some doing. We either need to enhance clientGetter somehow so I don't have to hack it like this or come up with some new kind of 'client getter' that can be used for my purposes, that supports both server-side and client-side (fake with unit tests)

Reply:
That's odd, in that if you look at the implementation of `extractToken()` just below your link, you'll see it at least intends to return an empty token with a nil error if there was no `authorization` in the context:

https://github.com/kubeapps/kubeapps/blob/master/cmd/kubeapps-apis/server/plugins.go#L353

I haven't checked, but perhaps `context.TODO()` doesn't include any metadata at all, and it's just that metadata is requried?

Reply:
In your case, you do actually want to use a context (as you're passing it to the `Watch` call below. So I'd not use `context.TODO()` but `context.Background()` below.

Reply:
this is raising an error in my usage 
```
	md, ok := metadata.FromIncomingContext(ctx)
	if !ok {
		return "", fmt.Errorf("error reading request metadata/headers")
	}

```

Reply:
Yep, or we can easily update the `extractToken()` to return an empty token with a nil error if there is no metadata with the context. That'd be cleaner IMO, and consistent with its current behaviour.

Reply:
ok, will make this change this as part of this PR